File: C-Plus-Plus/ds/BinarySearchTree.cpp
#include <iostream>
using namespace std;

/* A Binary Search Tree (BST) is a tree in which all the nodes follow the below-mentioned properties

The value of the key of the left sub-tree is less than the value of its parent (root) node's key.

The value of the key of the right sub-tree is greater than or equal to the value of its parent (root) node's key.*/

// This program implements insetion,Deletion and Searching in Binary Search Tree(BST)

struct Node{               //Structure is declared which holds Node data and address of left and right child
    struct Node *lchild;
    int data;
    struct Node *rchild;
}*root=NULL;      // root object is declared globally for struct Node

void insert(int element)    //Insertion happens here
{
    struct Node *p=root;   // using 2 pointers for keeping track of Nodes
    struct Node *q,*tmp;
    tmp= new Node;
    tmp->data=element;
    tmp->lchild=tmp->rchild=NULL;
    if(root==NULL)
    {
        root=tmp;
    }
    else
    {
        while(p!=NULL)
        {
            q=p;
            if(p->data==element)
            {
                return;
            }
            else if(element < p->data)  //if Inserted data is less than current Node data, move to left child
            {
                p=p->lchild;
            }
            else{
                p=p->rchild;    //if Inserted data is greater than or equal to current Node data, move to right child
            }
        }
        if(element > q->data) 
        {
            q->rchild=tmp;
        }
        else
            q->lchild=tmp;
    }
}

void Search(Node *p,int key)  //Seraching operation takes key as input and search for key in BST
{
    if(p==NULL)
    {
        cout << " BST is empty" << endl;
    }
    else{
        while(p!=NULL)
        {
            if(p->data==key)
            {
                cout << "Element is found" << endl;
                return;
            }
            else if(key < p->data)
            {
                p=p->lchild;
            }
            else{
                p=p->rchild;
            }
        }
    }
    cout << "Element not found "<<endl;
        
}

void InOrder(Node *p)   //Inorder travesal is used because inorder of BST gives Sorted output
{
    if(p)
    {
        InOrder(p->lchild);
        cout << p->data << " ";
        InOrder(p->rchild);
    }

}

int Height(Node *p)  //To find height of corresponding Node
{
int x,y;
if(p==NULL)return 0;
x=Height(p->lchild);
y=Height(p->rchild);
return x>y?x+1:y+1;
}

/*For deleting any Node on BST, Deleted Node has to be replaced with Appropiate Node without affecting the property of BST
If we are deleting any node, corresponding Node should be replaced by either InOrder predecessor or Inorder Successor based on height condition*/

struct Node * InOrderpredecssor(Node *p)  //If height of left child is greater than right child,it will enter this loop
{
    while(p && p->rchild!=NULL)
    {
        p=p->rchild;
    }
    return p;
}

struct Node * InOrderSuccessor(Node *p)  //If height of left child is less than right child,it will enter this loop
{
    while(p && p->lchild!=NULL)
    {
        p=p->lchild;
    }
    return p;
}

struct Node * Delete(Node *p,int key)
{
    struct Node *q;
    if(p==NULL)
    {
        return NULL;
    }
    if(p->lchild==NULL && p->rchild==NULL)
    {
        delete p;
        return NULL;
    }
    if(key < p->data)
    {
        p->lchild=Delete(p->lchild,key);
    }
    else if(key > p->data)
    {
        p->rchild=Delete(p->rchild,key);
    }
    else
    {
        if(Height(p->lchild) > Height(p->rchild))
        {
            q=InOrderpredecssor(p->lchild);
            p->data=q->data;
            p->lchild=Delete(p->lchild,q->data);
        }
        else
        {
            q=InOrderSuccessor(p->rchild);
            p->data=q->data;
            p->rchild=Delete(p->rchild,q->data);
            
        }
    }
return p;
}

int main()
{
    int Num,data,searchdata;
    cout << "Enter Number of data to be instered in BST: "<<endl;
    cin >> Num;
    cout <<"Enter the node values: " << endl;  //Enter the Node values to be inserted into BST
    for(int i=0;i<Num;i++)
    {
        cin >> data;
        insert(data);
    }
    cout <<"Elements in BST are: "<< endl;
    InOrder(root);
    cout << endl;
    cout << "Enter any node value to search in BST:" << endl;
    cin >> searchdata;
    Search(root,searchdata);
    cout <<"Enter the value to be deleted in BST: " << endl;
    cin >> data;
    Delete(root,data);
    cout << "After deleting Node: " << endl;
    InOrder(root);
    cout << endl;
    return 0;
}

File: C-Sharp/dp/BinarySearchTree.cs
/* Binary search tree(BST) is a rooted binary tree whose internal nodes each store a key greater than all 
the keys in the node's left subtree and less than those in its right subtree. */

using System;

namespace Binary_Search_Tree
{
    public class Node
    {
        public int data;
        public Node left;
        public Node right;
        public Node(int data)
        {
            this.data = data;
            left = null;
            right = null;
        }
    }
    public class BinarySearchTree
    {
        Node root;
        BinarySearchTree()
        {
            root = null;
        }
        public void Insert(int data)
        {
            root = Insert(root, data);
        }
        //Insert New Node in Binary search tree
        public Node Insert(Node root, int data)
        {
            if (root is null)
            {
                Node newNode = new Node(data);
                root = newNode;
                return root;
            }
            if (data < root.data)
            {
                root.left = Insert(root.left, data);
            }
            else if (data > root.data)
            {
                root.right = Insert(root.right, data);
            }
            return root;
        }
        //To limit parameters
        public void Inorder()
        {
            Inorder(root);
        }

        // to perform inorder traversal of BST
        public void Inorder(Node root)
        {
            if (root != null)
            {
                Inorder(root.left);
                Console.WriteLine(root.data);
                Inorder(root.right);
            }
        }
        //To limit parameters
        public void Postorder()
        {
            Postorder(root);
        }

        // to perform postorder traversal of BST
        public void Postorder(Node root)
        {
            if (root != null)
            {
                Postorder(root.left);
                Postorder(root.right);
                Console.WriteLine(root.data);
            }
        }
        //To limit parameters
        public void Preorder()
        {
            Preorder(root);
        }

        // to perform preorder traversal of BST
        public void Preorder(Node root)
        {
            if (root != null)
            {
                Console.WriteLine(root.data);
                Preorder(root.left);
                Preorder(root.right);
            }
        }
        //To limit parameters
        public int FindMin()
        {
            return FindMin(root);
        }
        //To find the min value in BST
        public int FindMin(Node root)
        {
            if(root.left is null)
            {
                return root.data;
            }
            return FindMin(root.left);
        }
        //To limit parameters
        public int FindMax()
        {
            return FindMax(root);
        }
        //To find the max value in BST
        public int FindMax(Node root)
        {
            if (root.right is null)
            {
                return root.data;
            }
            return FindMax(root.right);
        }
        //To search for a specific node in BST
        public Node Find(int data)
        {
            Node current = root;
            while (data != current.data)
            {
                if (data < current.data)
                {
                    current = current.left;
                }
                else
                {
                    current = current.right;
                }
                if (current is null)
                {
                    Console.WriteLine("Node is not found.");
                    return null;
                }
            }
            return current;
        }
        //To limit number of parameters
        public void Delete(int data)
        {
            root = Delete(root, data);
        }
        //To Delete a specific node
        public Node Delete(Node root, int data)
        {
            if (root == null)
                return root;
            //search for the node with the needed value
            if (data < root.data)
                root.left = Delete(root.left, data);
            else if (data > root.data)
                root.right = Delete(root.right, data);
            else
            {
                //for nodes with one or none child
                if (root.left == null)
                    return root.right;
                else if (root.right == null)
                    return root.left;
                //for nodes with two childs get min value in right
                root.data = FindMin(root.right);
                root.right = Delete(root.right, root.data);
            }
            return root;
        }
        public static void Main(String[] args)
        {
            BinarySearchTree BST = new BinarySearchTree();
            BST.Insert(50);
            BST.Insert(30);
            BST.Insert(20);
            BST.Insert(40);
            BST.Insert(70);
            BST.Insert(60);
            BST.Insert(80);

            Console.WriteLine("Inorder traversal");
            BST.Inorder();

            Console.WriteLine("Delete 20");
            BST.Delete(20);
            Console.WriteLine("Inorder traversal");
            BST.Inorder();
            Console.WriteLine("Preorder traversal");
            BST.Preorder();
            Console.WriteLine("Postorder traversal");
            BST.Postorder();
            Console.WriteLine("Min Value:");
            Console.WriteLine(BST.FindMin());
            Console.WriteLine("Max Value:");
            Console.WriteLine(BST.FindMax());



        }
    }
}
/*
Sample Input: 50 30 20 4070 60 80

inorder traversal: 20 30 40 50 60 70 80
Delete Node with data = 20
Inorder traversal: 30 40 50 60 70 80

Preorder traversal
50
30
40
70
60
80
Postorder traversal
40
30
60
80
70
50
Min Value:
30
Max Value:
80

Insersion Time Complexity: O(n)
Insersion Space Complexity: O(n)
Deletion Time Complexity: O(n)
Deletion Space Complexity: O(n)
Searching Time Complexity: O(n)
Searching Space Complexity: O(n)

*/
